### ISSUE TITLE
Envelope causes OOM with GeoHashPrefixTree

### ISSUE THREAD

**Issue by markwalkom** [elastic/elasticsearch#14181]
This looks like a bug, can someone check/confirm please :)

```
curl -X PUT http://localhost:9200/test/

curl -X PUT http://localhost:9200/test/_mapping/object -d '
{
    "object" : {
        "properties" : {
            "bbox" : {
                "type": "geo_shape",
                "precision": "1cm"
            }
        }
    }
}'

curl -X POST http://localhost:9200/test/object -d '
{
    "bbox" : {
        "type" : "envelope",
        "coordinates" : [ [-45.0, 45.0], [45.0, -45.0] ]
    }
}'
```

The logs from my attempt are pretty big so you can see them [here](https://gist.github.com/markwalkom/34a73229d888892fe005).

Reported initially [here](https://discuss.elastic.co/t/indexing-geo-shape-in-2-0-0-rc1-bug/32447).

**Comment by clintongormley** [comment]
1cm resolution for half the world?  Not surprising it OOMs :)

@nknize what can we do here?

**Comment by nknize** [comment]
With that high of resolution this uncovered an issue in Lucene's GeoHashPrefixTree that I had fixed in PackedQuadTree. I'd suggest cutting over to QuadTree with the following mappings:
 ```
        "properties" : {
            "bbox" : {
                "type": "geo_shape",
                "tree" : "quadtree",
                "precision": "29"
            }
```
This will give 1.11 cm accuracy but with shapes the size of the one posted I would ultimately suggest lowering that resolution, or at very least, breaking it down into smaller shapes.

**Comment by markwalkom** [comment]
/faceplam

I totally missed that and yes, makes sense

**Comment by nknize** [comment]
No worries Mark! It actually raised some good issues re: max precision. 1.  We throw a healthy IAE when ```precision``` is explicitly set beyond the max tree level but not when its set using "human readable" form (e.g., "1mm" happily self destructs w/ OOM). 2. GeoHashPrefixTree needs the same surgery procedure as applied to QuadPrefixTree

**Comment by michel-kraemer** [comment]
I agree that the resolution is too high. I initially used a smaller bounding box and got the same error. For the minimum working example to reproduce the bug I just copied the snippet from the docs. Sorry for the confusion.

So do I understand you correctly that there actually is a bug? Because the minimum example works quite well under 1.7. It's done within a couple of milliseconds even with the high resolution and the large bounding box.

Cheers,
Michel

**Comment by michel-kraemer** [comment]
OK. I see. I just tried it again with a smaller bounding box and it's working OK. It seems my application code produced too large bounding boxes and I didn't notice that. I will check my code. Thanks for the pointers.

**Comment by nknize** [comment]
I don't see this making 2.2 since it will be a lucene 5.5+ improvement. Removing version labels for now.

**Comment by michel-kraemer** [comment]
@nknize Now that we have Lucene 6, are there any plans to follow up on this issue?

**Comment by nknize** [comment]
closed in favor of #16749
